pages/learn/develop/container-contracts.md

```diff
@@ -0,0 +1,112 @@
+---
+title: Container contracts
+excerpt: A guide to using container contracts to enforce device compatibility
+---
+
+# Container contracts
+
+**Note** Container contracts are available for devices running Supervisor versions >= 10.6.17. All prior versions will not enforce any contracts.
+
+Container contracts are used to ensure the compatibility of a device to run an individual service of an application. For example, container contracts may be used to ensure that the device is running a specific version of [{{ $names.company.lower }} Supervisor][supervisor] or has a specific version of the [NVIDIA Tegra Linux Driver Package][l4t] (L4T).
+
+Each service container can define the contracts that it requires. If a container's contract requirements are not met, the release is not deployed to the device, unless it is marked as [optional](#optional-containers). You only need to define a contract for service containers that you wish to enforce a contract upon. Currently, you may define contract requirements based on the versions of Supervisor and L4T.
+
+## Contract specification
+
+Container contracts are defined as a `contract.yml` file and must be placed in the root of the build context for a service. For example, the following [multi-container application][multicontainer] defines two services named `first-service` and `second-service`, each with a corresponding `contract.yml` file.
+
+```shell
+.
+├── docker-compose.yml
+├── first-service
+│ ├── contract.yml
+│ └── Dockerfile.template
+└── second-service
+  ├── contract.yml
+  └── Dockerfile.template
+```
+
+`contract.yml` is a YAML file containing `type`, `slug`, `name`, and `requires` keys, which define the requirements of the contract. The following example requires the device to be running a version of L4T equal to 28.2 for the contract to pass. For L4T exact versions (either 28.2 or 32.2) must be specified.
```

Comment:
Could a use-case for optional containers be having one with a contract that depends on L4T 28.2 and one with a contract depending on L4T 32.2, and so therefore you'd always have one of the optional containers running on the device regardless of which version of L4T is installed? I know nothing about L4T... :laughing:

Reply:
I don't think so. L4T support is baked into the hostOS (https://github.com/balena-os/meta-tegra) so it's at the device level and not the container.

Reply:
Yeah, I guess I didn't explain that very well.
I meant if you had deviceA and deviceB with L4T 28.2 installed (in the hostOS) and deviceC and deviceD with L4T 32.2 installed, could you have optional container1 with a contract requiring L4T 28.2 and optional container2 with a contract requiring L4T 32.2; and then you could be sure that all devices would be using one-or-the-other of the optional containers?
(perhaps that's still not clear as I'm not very familiar with the current Balena terminology)